allow outways based on domain name

ronaldkoster opened the pull request:
In a (Delegated) Service Connection Grant an Outway can be specified based on domain name. Previously only a public key thumbprint was allowed. An Outway must have the domain name specified as a Subject Alternative Name in its certificate to be allowed to request an access token. By allowing domain names Contracts won't have to be renewed when the Outway renews its certificate and rotates its keypair.

TheBonheurs commented on specifications.md:
```diff
@@ -563,7 +563,8 @@ Before issuing an access token the Manager MUST validate that:
 1. The Manager is provided by a Peer who has an Inway which is offering the Service specified in `grant.data.service.name`. 
 1. The Peer ID specified by the X.509 certificate of the client requesting the access token matches the value of the field `grant.data.outway.peer_id`.
-1. The X.509 certificate provided by the client contains the same public key as specified in `grant.data.outway.public_key_fingerprint`
+1. The X.509 certificate provided by the client contains a public key with the same public key thumbprint  as specified in `grant.data.outway.public_key_thumbprint`. This validation should only be preformed when the value of `grant.outway.type` equals `OUTWAY_TYPE_PUBLIC_KEY_THUMBPRINT` 
```
```suggestion
1. The X.509 certificate provided by the client contains a public key with the same public key thumbprint  as specified in `grant.data.outway.public_key_thumbprint`. This validation should only be performed when the value of `grant.outway.type` equals `OUTWAY_TYPE_PUBLIC_KEY_THUMBPRINT` 
```

TheBonheurs commented on specifications.md:
```diff
@@ -563,7 +563,8 @@ Before issuing an access token the Manager MUST validate that:
 1. The Peer ID specified by the X.509 certificate of the client requesting the access token matches the value of the field `grant.data.outway.peer_id`.
-1. The X.509 certificate provided by the client contains the same public key as specified in `grant.data.outway.public_key_fingerprint`
+1. The X.509 certificate provided by the client contains a public key with the same public key thumbprint  as specified in `grant.data.outway.public_key_thumbprint`. This validation should only be preformed when the value of `grant.outway.type` equals `OUTWAY_TYPE_PUBLIC_KEY_THUMBPRINT` 
+1. The X.509 certificate provided by the client has a Subject Alternative Name(SAN) that matches the domain name specified in `grant.data.outway.domain_name`. This validation should only be preformed when the value of `grant.outway.type` equals `OUTWAY_TYPE_DOMAIN_NAME`  
```
```suggestion
1. The X.509 certificate provided by the client has a Subject Alternative Name (SAN) that matches the domain name specified in `grant.data.outway.domain_name`. This validation should only be preformed when the value of `grant.outway.type` equals `OUTWAY_TYPE_DOMAIN_NAME`  
```

TheBonheurs commented on media/specs/manager.yaml:
```diff
@@ -902,6 +937,12 @@ components:
       maxLength: 64
       minLength: 64
+    domainName:
+      description: A domain name according to RFC 1035
```
Are there any other restrictions on which domainnames can be used?

ronaldkoster replied:
I think the RFC and limits in the OAS are sufficient.

TheBonheurs commented on media/specs/manager.yaml:
```diff
@@ -688,17 +700,34 @@ components:
         - peer_id
         - public_key_thumbprint
+    outwayDomainName:
+      description: The details of the Outway based on a domain name for which a connection authorization was granted
```
```suggestion
      description: The details of the Outway based on a domain name for which a connection was authorized
```

I think this flows better but only if 'connection authorization' is not a defined term already.

TheBonheurs reviewed (commented):
Should we also add relevant test cases for this functionality (in the test-suite)?

RonaldKoster86 committed:
allow outways based on domain name

In a (Delegated) Service Connection Grant an Outway can be specified
based on domain name. Previously only a public key thumbprint was
allowed. An Outway must have the domain name specified as a Subject
Alternative Name in its certificate to be allowed to request an access
token. By allowing domain names Contracts won't have to be renewed when
the Outway renews its certificate and rotates its keypair.

Voles commented on specifications.md:
```diff
@@ -109,7 +109,10 @@ example Contract with a ServiceConnectionGrant
           "outway": {
             "peer_id": "00000000000000000002",
-            "public_key_thumbprint": "3a56f2e9269ac63f0d4394c46b96539da1625b6a985d38029ff89f34e490960c"
+            "identification" {
```
```suggestion
            "identification": {
```

Voles commented on specifications.md:
An example using a domain name is missing

TheBonheurs replied:
Discussed this with @ronaldkoster and concluded that adding an example is not needed. The referenced lines already show a Contract with a ServiceConnectionGrant, and there are no defined examples for OutwayIdentificationType in the standard beyond descriptive guidance. Adding such examples would also raise the question of adding examples for other types (e.g., ServiceTypes). 
We'll have to discuss this in a future release.

TheBonheurs committed:
missing colon

Co-authored-by: Niels Dequeker <niels@dqkr.be>
